### PR TITLE
API-1843: KMSEncryptionProvider Feature Gate

### PR DESCRIPTION
**Pull request by swghosh**
**Feature**: KMS Encryption Provider for sensitive etcd Resources

A user-configurable interface to support encryption of data stored in etcd using a supported Key Management Service (KMS).

OpenShift would need to align closer with KMS evolution upstream with respect to the different Kubernetes Encryption Providers available today that can encrypt resources from APIServer EncryptionConfig, https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/.

**User Stories**: 
- As an OpenShift administrator, I want to encrypt secrets in my cluster at rest using KMS keys so that I can comply with my organization's security requirements.
- As an OpenShift administrator, I want to let the KMS provider to manage the lifecycle of the encryption keys. 

xref: [OCPSTRAT-108](https://issues.redhat.com/browse/OCPSTRAT-108), [API-1684](https://issues.redhat.com/browse/API-1684)

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Please add an appropriate PR description to explain the motivation behind this change. Neither the PR, nor the card linked really explain what we are doing here, or why

**Comment by swghosh** [comment]
@JoelSpeed added description about the feature.
also, for context: #2035 
Not sure though why some of the e2e(s) are still failing.

**Comment by JoelSpeed** [comment]
According to the upstream documentation, this feature is stable as of 1.29, which would be 4.16.

It is my understanding that if an upstream feature is declared stable, and we have not otherwise specified the gate status, then the feature is already enabled.

This PR appears to disable a feature that has previously been enabled.

/hold

File: features/features.go [hunk 1]
```diff
@@ -662,4 +662,13 @@ var (
...
 								enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 								mustRegister()
+
+	FeatureGateKMSv2 = newFeatureGate("KMSv2").
+				reportProblemsToJiraComponent("kube-apiserver").
+				contactPerson("swghosh").
+				productScope(kubernetes).
+				enhancementPR("https://github.com/kubernetes/enhancements/issues/3299,https://github.com/openshift/enhancements/pull/1682").
+				enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
+				enableForClusterProfile(Hypershift, configv1.Default, configv1.TechPreviewNoUpgrade).
```
Comment (swghosh) [inline features/features.go:672]: HyperShift already uses upstream kube-apiserver KMSv2 (https://github.com/openshift/hypershift/pull/4502) so we need to enable this gate by default.

**Comment by swghosh** [comment]
> This PR appears to disable a feature that has previously been enabled.

Yes, that was the root cause causing those e2e failures `err="cannot set feature gate KMSv2 to false, feature is locked to true"`. `KMSv2` feature gate (https://github.com/kubernetes/enhancements/issues/3299) is enabled by default upstream and even [HyperShift uses it today](https://github.com/openshift/hypershift/pull/4502) so we do not have much gating to do on the upstream side. Today, in OpenShift we do not have the supported encryption providers to use the KMS v2 behaviour in downstream. Renaming the feature gate to `KMSEncryptionProvider` so we could gate our downstream implementation of the EncryptionConfig(s) through TP.

Thanks @JoelSpeed! 

xref: https://redhat-internal.slack.com/archives/C06DXMRLXQT/p1729766294150549

File: features/features.go [hunk 2]
```diff
@@ -662,4 +662,12 @@ var (
 								enhancementPR(legacyFeatureGateWithoutEnhancement).
 								enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 								mustRegister()
+
+	FeatureGateKMSv2 = newFeatureGate("KMSEncryptionProvider").
+				reportProblemsToJiraComponent("kube-apiserver").
+				contactPerson("swghosh").
+				productScope(kubernetes).
```
Comment (JoelSpeed) [inline features/features.go:669]: OpenShift scope not Kubernetes

Comment (swghosh) [reply features/features.go]: Fixed it.

File: features/features.go [hunk 3]
```diff
@@ -670,4 +670,11 @@ var (
 						enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 						enhancementPR("https://github.com/openshift/enhancements/pull/1697").
 						mustRegister()
+	FeatureGateKMSv2 = newFeatureGate("KMSEncryptionProvider").
```
Comment (JoelSpeed) [inline features/features.go:673]: Need a new line above this

Also, please be consistent about the naming
```suggestion
	
	FeatureGateKMSEncryptionProvider = newFeatureGate("KMSEncryptionProvider").
```

Comment (swghosh) [reply features/features.go]: thanks, fixed both (new line + naming).

**Comment by JoelSpeed** [comment]
PR is good to go, but we should get the EP into a state where the first round of reviews are in before we merge the gate, if we don't have at least some consensus, merging this would add a gate that is unneeded

**Comment by swghosh** [comment]
thanks @tkashem 
/remove-hold
off to you @JoelSpeed

**Comment by JoelSpeed** [comment]
/retest

Looks good, but I'd like to see if these test failures are related

**Comment by JoelSpeed** [comment]
/lgtm

PR is inert, I don't believe the tests failures are related and relevant components (eg KAS) don't recognise the gate, so we've got past my concern about overlapping gates

**Comment by JoelSpeed** [comment]
/hold It sounds like this feature might need a little more time to bake, I'd like to hold this until the API is ready to go and the EP has at least had some review.

From discussion with @wallylewis we don't need to rush this into 4.18 so lets wait until after branch cut

**Pull request merged**